### PR TITLE
MGS: Implement "/sp" endpoint to fetch state for all SPs

### PR DESCRIPTION
**Pull request by jgallagher**
There are three things I want to change after working on this PR:

1. I need to add some tests, at least of the `/sp` endpoint added here given its complexity. I'll do this before merging, but I think it's fine to start reviewing, particularly if there are any nontrivial changes that would affect those tests.
2. Error types are a bit of a mess and need some cleanup. I'll do this separately.
3. The way SPs are identified is all over the place; sometimes it's an `SpIdentifier`, sometimes its a `SocketAddr`, sometimes it's an ignition target (which itself is just an index, and is sometimes `u8` and sometimes `usize`). I've been punting on this until we have a bit better understanding of how MGS is going to interact with the management network and track rack topology, but it's pretty unwieldy at this point. I'll do this separately too.

There's also an open question of whether the SP communications should be separated out from `gateway` entirely so that they can be shared with RSS. I'm strongly inclined to do this (and try to at least make progress on items 2 and 3 above in doing so) even if RSS ends up calling MGS instead of communicating directly, just from a crate cleanliness/organization point of view. Thoughts welcome!

### CODE REVIEW

**Comment by jgallagher** [comment]
A couple basic tests are in place as of [90c49ca](https://github.com/oxidecomputer/omicron/pull/746/commits/90c49ca8e896143fb37f0813871de09110f45de8). I'd like to add more tests of some of the more complicated cases (e.g., an unresponsive SP), but that will require some more work in the SP simulator. May do that as part of this PR or as a followup, unsure at the moment.

**Comment by jgallagher** [comment]
Force pushed to account for #770

**Review by ahl — COMMENTED** [review]
great stuff!

File: gateway/src/http_entrypoints.rs [hunk 1]
```diff
@@ -254,10 +336,107 @@ pub(crate) struct PathSpComponent {
...
+        .into_iter()
+        .map(|BulkSpStateSingleResult { target, state, result }| {
+            let id = placeholder_map_from_target(known_sps, target)?;
+            let details = match result {
+                Ok(details) => details,
+                Err(SpCommsError::Timeout) => SpState::Unresponsive,
+                // TODO we're dropping the error on the floor here - how should
+                // we handle it? This is an SP that we actively failed to
+                // communicate with somehow, which isn't the same as
+                // "unresponsive". Should we fail the entire request? That's how
```
[the remark this answers is not in the document]
Comment (jgallagher) [reply gateway/src/http_entrypoints.rs:419]: The difficulty in answering this question is why I want to do some error cleanup! I can think of at least one other case that has to be handled, although in practice I don't expect to ever see it absent some horrible deployment mismatch nightmare: the SP sends a non-error response of a type that doesn't make sense (e.g., we ask it for its state and it responds with "here's a list of my components").

File: gateway/src/sp_comms/bulk_state_get.rs [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+// Copyright 2022 Oxide Computer Company
+
+//! Implementation to fetch the state of all powered-on SPs, with support for
+//! pagination across shorter timeouts than the overall timeout to collect all
+//! state.
+//!
+//! The flow of this process, starting from the client(s), is below. This
+//! modules comes into play at step 3.
+//!
+//! 1. A client requests the state of all SPs. They provide a timeout for the
```
[the remark this answers is not in the document]
Comment (jgallagher) [reply gateway/src/sp_comms/bulk_state_get.rs:14]: changing to `The client provides...`

File: gateway/src/sp_comms/bulk_state_get.rs [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+                    }
+                }
+
+                if skip_results == 0 {
+                    // caller claimed to have seen a target, but we don't have
+                    // it (if we did, `skip_results` is >= 1).
+                    return Err(Error::InvalidLastSpSeen);
+                }
+
+                // go ahead and check to see if we already have enough info to
```
[the remark this answers is not in the document]
Comment (jgallagher) [reply gateway/src/sp_comms/bulk_state_get.rs:245]: I'm gonna go ahead and take that as a compliment, bless your heart

**Pull request merged**